src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/optional.hpp

```diff
@@ -14,93 +14,677 @@
+            return false;
+        }
+        return !left.has_value() || *left == *right;
+    }
+
+    template <typename T, typename U>
+    friend constexpr auto tag_invoke(bsoncxx::detail::equal_to,
+                                     optional<T> const& left,
+                                     U const& right) noexcept -> bsoncxx::detail::
+        requires_t<bool, not_an_optional<U>, bsoncxx::detail::is_equality_comparable<T, U>> {
+        return left.has_value() && *left == right;
+    }
+
+    template <typename T>
+    friend constexpr bool tag_invoke(bsoncxx::detail::equal_to,
+                                     optional<T> const& opt,
+                                     nullopt_t) noexcept {
+        return !opt.has_value();
+    }
+
+    template <typename T, typename U>
+    bsoncxx_cxx14_constexpr friend auto tag_invoke(bsoncxx::detail::compare_three_way compare,
+                                                   optional<T> const& left,
+                                                   optional<U> const& right)
+        -> bsoncxx::detail::requires_t<bsoncxx::detail::strong_ordering,
+                                       bsoncxx::detail::is_totally_ordered_with<T, U>> {
+        if (left.has_value()) {
+            if (right.has_value()) {
+                return compare(*left, *right);
+            } else {
+                // non-null is greater than any null
+                return bsoncxx::detail::strong_ordering::greater;
+            }
+        } else {
+            if (right.has_value()) {
+                // Null is less than any non-null
+                return bsoncxx::detail::strong_ordering::less;
+            } else {
+                // Both are null
+                return bsoncxx::detail::strong_ordering::equal;
+            }
+        }
+    }
+
+    template <typename T, typename U>
+    bsoncxx_cxx14_constexpr friend auto tag_invoke(bsoncxx::detail::compare_three_way compare,
+                                                   optional<T> const& left,
+                                                   U const& right)
+        -> bsoncxx::detail::requires_t<bsoncxx::detail::strong_ordering,
+                                       not_an_optional<U>,
+                                       bsoncxx::detail::is_totally_ordered_with<T, U>> {
+        if (left.has_value()) {
+            return compare(*left, right);
+        }
+        // null optional is less-than any non-null value
+        return bsoncxx::detail::strong_ordering::less;
+    }
+
+    template <typename T>
+    constexpr friend bsoncxx::detail::strong_ordering tag_invoke(bsoncxx::detail::compare_three_way,
+                                                                 optional<T> const& left,
+                                                                 nullopt_t) {
+        return left.has_value() ? bsoncxx::detail::strong_ordering::greater
+                                : bsoncxx::detail::strong_ordering::equal;
+    }
+};
```

Comment:
```suggestion
    template <typename T>
    constexpr friend bsoncxx::detail::strong_ordering tag_invoke(bsoncxx::detail::compare_three_way compare,
                                                                 optional<T> const& left,
                                                                 nullopt_t) {
        return compare(left.has_value(), false);
    }
```
Relational comparison operators do not yield the correct result when `nullopt` is the left-hand argument and the optional `x` contains a value:

| Op | bool(x) | Correct | Actual
| -- | ------- | ------- | ------
| == | false   | true    | true
| != | false   | false   | false
| <  | false   | false   | false
| <= | false   | true    | true
| >  | false   | false   | false
| >= | false   | true    | true
| == | true    | false   | false
| != | true    | true    | true
| <  | true    | true    | false (!)
| <= | true    | true    | false (!)
| >  | true    | false   | true (!)
| >= | true    | false   | true (!)

Behavior seems correct when `nullopt` is the right-hand argument or when the optional does not contain a value.

C++20 spec suggests the definition of the three-way comparison operator should be `bool(x) <=> false` as suggested above, but this does not seem to fix the problem. Is there perhaps a bug in the implementation of `bsoncxx::detail::compare_three_way`?

Reply:
The issue was with incorrect logic in the argument-swapped hidden-friend operators. Fixed.